File: bridge/fallback.py
"""Explicit fallback, identical to the Go plugin's `FallbackName`.

It must never raise and must not depend on live telemetry: it runs precisely when
something else already went wrong. A fallback is a decision of the method under test, so
it is logged, counted, and kept in the results.

The second criterion matters: in the extender path per-node requested resources are
unknown, so every free ratio is 1 and the choice would otherwise collapse to name order.
"""

NAME = "least-allocated-requests"


def fallback_node(snapshot: dict) -> str:
    nodes = snapshot["nodes"]
    if not nodes:
        return ""
    ranked = sorted(nodes, key=lambda n: (-_free_ratio(n), -n["allocatable_millicpu"], n["name"]))
    return ranked[0]["name"]


def fallback_scores(snapshot: dict) -> dict[str, float]:
    chosen = fallback_node(snapshot)
    return {n["name"]: (1.0 if n["name"] == chosen else 0.0) for n in snapshot["nodes"]}


def _free_ratio(node: dict) -> float:
    allocatable = node["allocatable_millicpu"]
    if allocatable <= 0:
        return 0.0
    free = allocatable - node["requested_millicpu"]
    return free / allocatable if free > 0 else 0.0

File: plugin/pkg/kaptain/fallback.go
package kaptain

import (
	"sort"

	"github.com/leopoldch/kaptain/plugin/pkg/snapshot"
)

// FallbackName runs when something else already failed, so it uses no telemetry of its own.
// A fallback is a decision of the method under test: logged, counted, kept in the results.
const FallbackName = "least-allocated-requests"

func fallbackNode(s *snapshot.Snapshot) string {
	nodes := append([]snapshot.Node(nil), s.Nodes...)
	sort.Slice(nodes, func(i, j int) bool {
		left, right := nodes[i], nodes[j]
		if free := freeRatio(left); free != freeRatio(right) {
			return free > freeRatio(right)
		}
		if left.AllocatableMilliCPU != right.AllocatableMilliCPU {
			return left.AllocatableMilliCPU > right.AllocatableMilliCPU
		}
		return left.Name < right.Name
	})
	if len(nodes) == 0 {
		return ""
	}
	return nodes[0].Name
}

func fallbackScores(s *snapshot.Snapshot) map[string]float64 {
	chosen := fallbackNode(s)
	scores := make(map[string]float64, len(s.Nodes))
	for _, n := range s.Nodes {
		scores[n.Name] = 0
	}
	if chosen != "" {
		scores[chosen] = 1
	}
	return scores
}

func freeRatio(n snapshot.Node) float64 {
	if n.AllocatableMilliCPU <= 0 {
		return 0
	}
	free := n.AllocatableMilliCPU - n.RequestedMilliCPU
	if free <= 0 {
		return 0
	}
	return float64(free) / float64(n.AllocatableMilliCPU)
}
